src/electionguard/voting/coordinator.c

```diff
@@ -8,221 +8,318 @@
 #include "sha2-openbsd.h"
+#include "voting/ballot_collection.h"
 #include "voting/message_reps.h"
-#include "voting/num_ballots.h"
 
-// @design jwaksbaum This implementation relies on the fact that the
-// Encrypters are generating uids from a shared counter, so we can
-// just store the ballots at the index of their uid and they will all
-// be contiguous. In a real implementation you would probably need a
-// hash table.
+// @design mwilhelm This implementation utilizes a hash table to keep
+// track of external ballot identifiers (strings) that are already
+// registered, cast, or spoiled.  Additionally, to support low-memory
+// systems it also allows batch-processing when importing ballots.
+// there should only ever be one voting coordinator in memory on a system
+// and it should retain it's state through the ballot load and cast cycle
 
+/**
+ * The current state of a voting coordinator
+ */
 struct Voting_Coordinator_s
 {
-    // The number of selections on each ballot
+    // The number of selections available on each ballot
     uint32_t num_selections;
-    struct encryption_rep *selections[MAX_BALLOTS];
-    bool registered[MAX_BALLOTS];
-    bool cast[MAX_BALLOTS];
-    bool spoiled[MAX_BALLOTS];
-    char *tracker[MAX_BALLOTS];
+
+    // count of ballots registered
+    uint32_t registered_num_ballots;
+
+    // count of ballots in the buffer
+    uint32_t buffered_num_ballots;
+
+    // external id buffer
+    char *buffered_external_id[MAX_BALLOT_PAYLOAD];
+
+    // selections buffer
+    struct encryption_rep *selections[MAX_BALLOT_PAYLOAD];
 };
 
+static int Voting_Coordinator_ref_count = 0;
+
+// use a static result as a singleton instance
+// since only one coordinator should exist
+// per application instance
+static struct Voting_Coordinator_new_r voting_coordinator_singleton;
+
 struct Voting_Coordinator_new_r Voting_Coordinator_new(uint32_t num_selections)
 {
-    struct Voting_Coordinator_new_r result;
-    result.status = VOTING_COORDINATOR_SUCCESS;
+    voting_coordinator_singleton.status = VOTING_COORDINATOR_SUCCESS;
 
-    // Allocate the ballot box
-    result.coordinator = malloc(sizeof(struct Voting_Coordinator_s));
-    if (result.coordinator == NULL)
-        result.status = VOTING_COORDINATOR_INSUFFICIENT_MEMORY;
+    // sanity-check if the instance exists and matches this election configuration
+    if (voting_coordinator_singleton.coordinator != NULL 
+        && voting_coordinator_singleton.coordinator->num_selections == num_selections)
+    {
+#ifdef DEBUG_PRINT
+        printf("\nVoting_Coordinator_new: already exists.  Returning existing instance\n");
+#endif
+        Voting_Coordinator_ref_count++;
+        return voting_coordinator_singleton;
+    }
 
-    // Initialize the ballot box
-    if (result.status == VOTING_COORDINATOR_SUCCESS)
+    if (voting_coordinator_singleton.coordinator != NULL 
+        && voting_coordinator_singleton.coordinator->num_selections != num_selections)
```

Comment:
it may be a valid use case that the selections do not match, particularly when supporting multiple ballot styles

Reply:
#76